### PR TITLE
(#4515) Add Events Dynamic Lists MLP

### LINKED ISSUE #4515: Story: Add Events Dynamic List to MLP
### User Needs
- 3.23: As a drupal user of an MLP, I need to have components/features that allow for updating information automatically.

### Requirements

**Front-End**
Figma: [Comps](https://www.figma.com/design/qVjky6CIVCeompBkkuIuUy/Inner-Pages?node-id=271-11328&t=vUH6zML8eEJPea0Q-1)

- This will cover both the Past and Future events lists (this will most likely be 2 Views for the listing of items in past and future)
  - Both past and upcoming events pages will indicate the number of events displayed on the current page if there are more events in the respective grouping than are shown on the individual page (in the form of "Past Events X - Y of Z". E.g, "Past Events 1-10 of 50"). The pager and this text will be displayed at the bottom of their respective pages.
- Events do not have their URL changed when they go from future to past
- Event Series label, address (venue, room, city/state/country) optional
- Title, date, time, list description required
- Base the front end display off Collection Default. 
  - Replace Name with Location
  - Date needs to be able to add time
  - Description = List Description
  - Need to add area for Event Series Label
  - Will need to edit macro to accomplish these changes
  - The order of the fields for an event type will be changed from our Collections Default to match the current view of Events (see link [here](https://dceg.cancer.gov/news-events/events/past-events) for example)
  - Will need pagination (use USWDS pagination Default)
- See related ticket #3614, #3462, and #2289 

**CMS**
- Requirement: Events shall be able to be filtered by Event Series so that all events in an instance of the CMS do not always appear within every list of events. (Need to ensure that the creation of events does not get shown on every page that uses events.)
- It will reuse the "NCIDS Dynamic List" paragraph from #4516, and it will remain able to be added to the Contents field on Mini Landing Pages
- It will be done by creating two new Views which will be called "NCIDS Collection Events Past" and "NCIDS Collection Events Future", and they will be the listing of items for the Dynamic List
- It will only function on the NCIDS Style template on MLPs
- Update the NCIDS Dynamic List paragraph so that the "Source View" field is required. 

### DoR Tasks

- [x] Content Model Drafted
- [x] Images attached to ticket (Design)
- [x] Content Model Reviewed
  - [x] Developer Review
  - [x] IA Review
- [x] Content Model Approved
  - [x]  Developer Approval
  - [x] IA Approval
- [x] Acceptance Criteria Drafted
  - [x] Analytics Acceptance Criteria Drafted (As applicable)
- [x] Scenarios created for YAML
- [x] Acceptance Criteria Reviewed
  - [x] Developer Review
  - [x] IA Review
- [x] Acceptance Criteria Approved
  - [x] Developer Approval
  - [x] IA Approval
- [x] Solve the TBD determination in CMS specifications above: Result - use the Events Series

### DoD Tasks

- [ ] Governance Drafted
- [ ] PR Review (see PR for review and comments)
- [ ] Client Review with Demo

### Technical Notes

**Content Model:** [Content Model](https://github.com/NCIOCPL/cgov-digital-platform/wiki/Paragraph-Type-NCIDS-Dynamic-List) this will be reused from the NCIDS Dynamic List used on the News & Events page, and the Press Release Dynamic List added on MLPs.

**Scenarios/Comps**
- To test the Past Events page, the [DCEG Past Events page](https://dceg.cancer.gov/news-events/events/past-events) will be recreated on a Mini Landing Page by creating a new MLP using the NCIDS Dynamic List.
  - The page title will be "Past Events"
  - There will be a section added to the top of the Past Events view that displays the number of events shown on the current page out of the total number of past events (e.g., "Past Events 1-10 of 50")
  - The dynamic list will used the [Collection Default](https://designsystem-dev.cancer.gov/develop/storybook/?path=/story/components-collection-default--default), but the name will be replaced by the location, the date will add a time, and the description will be the event's list description
  - The [Default Pagination](https://designsystem-dev.cancer.gov/develop/storybook/?path=/story/uswds-components-pagination--default) will be added to the bottom of the page
  - Create the site section `/news-events/events/past-events`
- To test the Upcoming Events page, the [DCEG Past Events page](https://dceg.cancer.gov/news-events/events/past-events) will be recreated with altered dates by recreating it on a Mini Landing Page by creating a new MLP using the NCIDS Dynamic List. (NOTE: We may need to get creative with this as all upcoming events are cancelled for the time being, so there is no reference for testing)
  - The page title will be "Upcoming Events"
  - The dynamic list will use the [Collection Default](https://designsystem-dev.cancer.gov/develop/storybook/?path=/story/components-collection-default--default), but the name will be replaced by the location, the date will add a time, and the description will be the event's list description
  - The [Default Pagination](https://designsystem-dev.cancer.gov/develop/storybook/?path=/story/uswds-components-pagination--default) will be added to the bottom of the page
  - Create the site section `/news-events/events/upcoming-events`

### Acceptance Criteria

**CMS AC**
```gherkin
Scenario: A content editor for the Content Management System (CMS) adds an NCIDS Dynamic List to a Mini Landing Page (MLP)
  Given a user is logged into the CMS as an editor
    And they are editing a Mini Landing Page
    And the NCIDS Dynamic List has been added as a paragraph to the Mini Landing Page
    And the NCIDS Dynamic List functions on an MLP with the NCIDS Style
  When the user is editing an MLP
    And they click the dropdown arrow to add paragraphs under "Contents"
  Then the "Add NCIDS Dynamic List" option will appear as an option on the dropdown
  When the user clicks the "Add NCIDS Dynamic List" button
  Then the fields for the NCIDS Dynamic List paragraph will appear (see the link to the content model above) in the Contents field
  When the user fills out the required fields of the NCIDS Dynamic List
    And they select the Source View field option 'NCIDS Collection Events Past'
    And any optional fields they desire
    And they save the page
  Then the page will be updated with the Dynamic List using the NCIDS Collections Events Past view
  When the user fills out the required fields of the NCIDS Dynamic List
    And they select the source view field option "NCIDS Collection Events Future" view
    And any optional fields they desire
    And they save the page
  Then the page will be updated with the Dynamic List using the NCIDS Collections Events Future view
  When the user does not fill out all the required fields of the NCIDS Dynamic List
    And they save the page
  Then they will receive an error indicating that the required fields are not filled out
```

**Front-End AC**
```gherkin
Scenario: A user views a Mini Landing Page with an NCIDS Dynamic List with NCIDS Collections Events Past or NCIDS Collections Events Future view added to it
  Given the NCIDS Dynamic List with the has been added to the Mini Landing Page
    And the NCIDS Dynamic List uses the NCIDS Collections Events Past or NCIDS Collections Events Future view
  When the user views the NCIDS Dynamic List on the MLP
  Then it will appear with the usa-collection style
    And the titles of the list will appear as links
    And the name item will be replaced with the location
    And the date item will display the date and time
    And the description will be the list description of the event
  When the user hovers their mouse over the title of the item
  Then the hover state color of the title will appear
  When the user clicks on the title of the item
  Then the user will be redirected to the page linked in the item
```

### PR DESCRIPTION
**Pull request by erikscottmeyer**
(Closes #4515).

### CODE REVIEW

**Comment by erikscottmeyer** [comment]
Test URLs from Yaml:
/news-events/events/past-events-collection
/news-events/events/future-events-collection
/news-events/events/past-events-collection-test-nci-event-series
/news-events/events/future-events-collection-test-nci-event-series
/news-events/events/past-events-collection-test-pager
/news-events/events/future-events-collection-test-pager

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/templates/listing/list_item_macros.twig [hunk 1]
```diff
@@ -339,3 +339,66 @@
...
+            {% endif %}
+          {% else %}
+            {{ eventStart|date('F j, Y') }}
+            {% if allDay != "On" %}
+              &vert;
+              {{ eventStart|date('g:i A') }}
+              &ndash;
+              {{ eventEnd|date('g:i A') }}
+            {% endif %}
+          {% endif %}
```
Comment (bryanpizzillo) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/templates/listing/list_item_macros.twig:382]: These should probably be `<time>` elements in keeping with our guidance on using dates and times as `<time>` elements. 

Also, there is no real context here -- this is something schema.org has a schema for, so adding context is probably for the future.

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/templates/content/event/views-view--cgov-events--ncids_collection_condensed_past_events.html.twig [hunk 2]
```diff
@@ -0,0 +1,16 @@
+<ul class="usa-collection usa-collection--condensed" data-dynamic-list-view="Future Events" data-dynamic-list-display="NCIDS Collection Condensed">
```
Comment (bryanpizzillo) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/templates/content/event/views-view--cgov-events--ncids_collection_condensed_past_events.html.twig:1]: The `data-dynamic-list-view` should probably be something other than "Future Events".

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/templates/content/event/views-view--cgov-events--ncids_collection_condensed_past_events.html.twig [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    {% if rows -%}
+      {{ rows }}
```
Comment (bryanpizzillo) [inline docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/templates/content/event/views-view--cgov-events--ncids_collection_condensed_past_events.html.twig:3]: Probably the same `views-row` comment as above here.

**Review by bryanpizzillo — CHANGES_REQUESTED** [review]
Some notes below within code. Also, can you please make the future events in the year `2050` or something? This Thursday is only the future until, well, this Thursday.

Also, `/news-events/events/future-events-collection-test-pager` has no pager showing. 

Additionally, the future events should have a mix of events with descriptions and without.

**Comment by bryanpizzillo** [comment]
@erikscottmeyer can you also add the Javascript for turning on Dynamic List analytics to `docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/entrypoints/ncids-mini-landing/ncids-mini-landing.ts`?

This should be pulling in `cgdpDynamicListInit` just like on https://github.com/NCIOCPL/cgov-digital-platform/blob/develop/docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/entrypoints/ncids-home-landing/ncids-home-landing.ts

File: docroot/profiles/custom/cgov_site/modules/custom/cgov_list/config/install/field.field.paragraph.ncids_dynamic_list.field_source_view.yml [hunk 4]
```diff
@@ -12,7 +12,7 @@ entity_type: paragraph
 bundle: ncids_dynamic_list
 label: 'Source View'
 description: 'The source Drupal View for the items to display.'
-required: false
+required: true
```
[the remark this answers is not in the document]
Comment (erikscottmeyer) [reply docroot/profiles/custom/cgov_site/modules/custom/cgov_list/config/install/field.field.paragraph.ncids_dynamic_list.field_source_view.yml:15]: This was a requirement in the Ticket:
Update the NCIDS Dynamic List paragraph so that the "Source View" field is required.

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/templates/listing/list_item_macros.twig (hunk 1 above)
Comment (bryanpizzillo) [reply docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/templates/listing/list_item_macros.twig:382]: We should skip time for now. Really we should make sure the date is honoring timezones and whatnot for this to be useful for computers.

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/templates/content/event/views-view--cgov-events--ncids_collection_condensed_future_events.html.twig [hunk 5]
```diff
@@ -0,0 +1,17 @@
+<ul class="usa-collection usa-collection--condensed" data-dynamic-list-view="Future Events" data-dynamic-list-display="NCIDS Collection Condensed">
+    {% if rows -%}
+      {{ rows }}
```
[the remark this answers is not in the document]
Comment (erikscottmeyer) [reply docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/templates/content/event/views-view--cgov-events--ncids_collection_condensed_future_events.html.twig:3]: These are set in the Views UI as discussed (Format -> Settings -> Checkbox - Add views row classes.  It's checked on by default.

[the remark this answers is not in the document]
Comment (erikscottmeyer) [reply docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/templates/content/event/views-view--cgov-events--ncids_collection_condensed_future_events.html.twig:3]: Right, this can be set in the view UI, but that leaves an empty div.  It was removed completely in NCIDS Collection Media by adding a new Template views-view-unformatted--press-releases--ncids_collection_media.html.twig ...  {% for row in rows %}
  {{- row.content -}}
{% endfor %}
So I added the equivalent for these.

File: docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/templates/listing/paragraph--ncids-dynamic-list.html.twig [hunk 6]
```diff
@@ -6,7 +6,7 @@
   other fields, we might as well handle the title as well.
 #}
 
-<div class="cgdp-dynamic-list" data-eddl-landing-row>
+<div class="cgdp-dynamic-list usa-section" data-eddl-landing-row>
```
[the remark this answers is not in the document]
Comment (erikscottmeyer) [reply docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/templates/listing/paragraph--ncids-dynamic-list.html.twig]: I updated the YAML for Press Releases (310_press_archive.content.yml) to wrpa the raw html in <div class="usa-section">

**Comment by welshja** [comment]
Analytics Testing:
- On https://ncigovcdode922.prod.acquia-sites.com/news-events/events/past-events-collection-test-nci-event-series and all the analytics data values are coming through except I'm seeing ERROR for the event.detail.data.title value. This gets populated by the "List Title" field.
- On https://ncigovcdode922.prod.acquia-sites.com/news-events/events/future-events-collection-test-nci-event-series, I'm getting the "List Title" value in the event.detail.data.title value, which is good.
- Seeing the same ERROR value on https://ncigovcdode922.prod.acquia-sites.com/news-events/press-releases/2018 where there is no List Title value shown.

Can we please update these analytics so that if the "List Title" field is left blank that the value is set to "No Title"? Thanks.

@erikscottmeyer @andyvanavery31

**Comment by erikscottmeyer** [comment]
@welshja I've made the change you requested and pushed.

**Review by andyvanavery31 — APPROVED** [review]

**Comment by KateMashkinaNIH** [comment]
Analytics for event link click is passing a wrong total number of links when there is a pager - it counts the event links + the pager links, these two should be separated

**Comment by bryanpizzillo** [comment]
Just adding the comment from Slack from Joe:
> after kicking all this around with the pager on events and press releases, let's go with it as is with the pager items getting the same click event as everything else in the list. the issue Kate identified with the total numbers of links is one we should accept for now.
> Landing on this for a few reasons:
the ticket really didn't specify anything related to this or analytics changes needed, so adding some requirements to rework this now feels obnoxious to rework to how I would request it would require reworking of a bunch of regression tests (which Kate is out, so would be blocked from)
> it is a bit of an edge case in terms of the places we will be using this component on [c.gov](http://c.gov/), so lower risk
I'll create a change request now to address in the future. It will request to create a new click event specifically for pager and view more (and stop counting those as part of the items in the list).

File: docroot/profiles/custom/cgov_site/modules/custom/cgov_yaml_content/content/310_press_archive.content.yml [hunk 7]
```diff
@@ -39,12 +39,12 @@
       field_raw_html:
         - format: "raw_html"
           value: |
-            <p><a href="/news-events/press-releases/2025">2025</a>  |&nbsp;
+            <div class="usa-section"><p><a href="/news-events/press-releases/2025">2025</a>  |&nbsp;
```
Comment (bryanpizzillo) [inline docroot/profiles/custom/cgov_site/modules/custom/cgov_yaml_content/content/310_press_archive.content.yml:42]: This should be `<div class="usa-section" data-eddl-landing-row="">` so that the analytics for the mini landing count it. 

`data-eddl-landing` is needed for each section of a landing page.

File: docroot/profiles/custom/cgov_site/modules/custom/cgov_yaml_content/content/310_press_archive.content.yml [hunk 8]
```diff
@@ -171,7 +171,7 @@
       field_raw_html:
         - format: "raw_html"
           value: |
-            <p>2018 | <a href="/espanol/noticias/comunicados-de-prensa/2017" title="">2017</a> | <a href="/espanol/noticias/comunicados-de-prensa/2016" title="">2016</a> | <a href="/espanol/noticias/comunicados-de-prensa/2015" title="">2015</a> | <a href="/espanol/noticias/comunicados-de-prensa/2014" title="">2014</a></p>
+            <div class="usa-section"><p>2018 | <a href="/espanol/noticias/comunicados-de-prensa/2017" title="">2017</a> | <a href="/espanol/noticias/comunicados-de-prensa/2016" title="">2016</a> | <a href="/espanol/noticias/comunicados-de-prensa/2015" title="">2015</a> | <a href="/espanol/noticias/comunicados-de-prensa/2014" title="">2014</a></p></div>
```
Comment (bryanpizzillo) [inline docroot/profiles/custom/cgov_site/modules/custom/cgov_yaml_content/content/310_press_archive.content.yml:174]: Same

**Review by bryanpizzillo — CHANGES_REQUESTED** [review]
One small change to the raw html blocks so that they get the data attribute that the JS looks at to get the total number of items on a page.

File: docroot/profiles/custom/cgov_site/modules/custom/cgov_list/config/install/field.field.paragraph.ncids_dynamic_list.field_source_view.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (bryanpizzillo) [reply docroot/profiles/custom/cgov_site/modules/custom/cgov_list/config/install/field.field.paragraph.ncids_dynamic_list.field_source_view.yml:15]: Okie dokie.

**Comment by welshja** [comment]
Reverted analytics for Events MLP looking good to me on the ODE
